producer/writer/message_writer.go

```diff
@@ -279,27 +290,41 @@ func (w *messageWriterImpl) retryUnacknowledged() {
 	var (
 		toBeRetried []*message
 		beforeRetry = w.nowFn()
+		batchSize   = w.opts.MessageRetryBatchSize()
 	)
 	for e != nil {
-		now := w.nowFn()
-		nowNanos := now.UnixNano()
+		beforeBatch := w.nowFn()
+		beforeBatchNanos := beforeBatch.UnixNano()
 		w.Lock()
-		e, toBeRetried = w.retryBatchWithLock(e, nowNanos)
+		e, toBeRetried = w.retryBatchWithLock(e, beforeBatchNanos, batchSize)
 		consumerWriters := w.consumerWriters
 		w.Unlock()
-		if len(consumerWriters) == 0 {
-			// Not expected in a healthy/valid placement.
-			w.m.noWritersError.Inc(int64(len(toBeRetried)))
-			w.m.retryBatchLatency.Record(w.nowFn().Sub(now))
-			continue
+		err := w.writeBatch(consumerWriters, toBeRetried)
+		w.m.retryBatchLatency.Record(w.nowFn().Sub(beforeBatch))
+		if err != nil {
```

Comment:
Do you need to check the type of error here to ensure you avoid retry if the error is what's expected?

Reply:
we only return err when we can't write to any consumer, it's like a boolean so no need to check